File: src/main/java/com/paragon/client/ui/panel/element/setting/BindElement.java
package com.paragon.client.ui.panel.element.setting;

import com.paragon.Paragon;
import com.paragon.api.event.client.SettingUpdateEvent;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.api.setting.Bind;
import com.paragon.api.setting.Setting;
import com.paragon.client.ui.panel.Click;
import com.paragon.client.ui.panel.element.Element;
import com.paragon.client.ui.panel.element.module.ModuleElement;
import net.minecraft.util.math.MathHelper;
import org.lwjgl.input.Keyboard;

import java.awt.Color;

@SuppressWarnings("unchecked")
public final class BindElement extends Element {

    private final Setting<Bind> setting;

    private boolean listening;
    private float hover;
    private float enabled;

    public BindElement(int layer, Setting<Bind> setting, ModuleElement moduleElement, float x, float y, float width, float height) {
        super(layer, x, y, width, height);

        setParent(moduleElement.getParent());
        this.setting = setting;

        setting.getSubsettings().forEach(subsetting -> {
            if (subsetting.getValue() instanceof Boolean) {
                getSubElements().add(new BooleanElement(layer + 1, (Setting<Boolean>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            } else if (subsetting.getValue() instanceof Enum<?>) {
                getSubElements().add(new EnumElement(layer + 1, (Setting<Enum<?>>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            } else if (subsetting.getValue() instanceof Number) {
                getSubElements().add(new SliderElement(layer + 1, (Setting<Number>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            } else if (subsetting.getValue() instanceof Bind) {
                getSubElements().add(new BindElement(layer + 1, (Setting<Bind>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            } else if (subsetting.getValue() instanceof Color) {
                getSubElements().add(new ColourElement(layer + 1, (Setting<Color>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            } else if (subsetting.getValue() instanceof String) {
                getSubElements().add(new StringElement(layer + 1, (Setting<String>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }
        });
    }

    @Override
    public void render(int mouseX, int mouseY, int dWheel) {
        if (setting.isVisible()) {
            hover = MathHelper.clamp(hover + (isHovered(mouseX, mouseY) ? 0.02f : -0.02f), 0, 1);
            enabled = MathHelper.clamp(enabled + (listening ? 0.02f : -0.02f), 0, 1);

            RenderUtil.drawRect(getX(), getY(), getWidth(), getHeight(), new Color(40, 40, 45).getRGB());
            RenderUtil.drawRect(getX() + getLayer(), getY(), getWidth() - getLayer() * 2, getHeight(), new Color((int) (40 + (30 * hover)), (int) (40 + (30 * hover)), (int) (45 + (30 * hover))).getRGB());
            RenderUtil.drawRect(getX() + getLayer(), getY(), 1, getHeight() * enabled, Color.HSBtoRGB(getParent().getLeftHue() / 360, 1, 0.5f + (0.25f * hover)));

            renderText(setting.getName(), getX() + (getLayer() * 2) + 5, getY() + getHeight() / 2 - 3.5f, 0xFFFFFFFF);
            renderText(setting.getValue().getButtonName(), (getX() + getWidth() - (getLayer() * 2)) - getStringWidth(setting.getValue().getButtonName()) - 3, getY() + getHeight() / 2 - 3.5f, new Color(150, 150, 155).getRGB());

            super.render(mouseX, mouseY, dWheel);
        }
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, Click click) {
        if (setting.isVisible()) {
            if (isHovered(mouseX, mouseY) && getParent().isElementVisible(this) && click.equals(Click.LEFT)) {
                listening = !listening;
                return;
            }

            if (listening) {
                listening = false;

                setting.getValue().setDevice(Bind.Device.MOUSE);
                setting.getValue().setButtonCode(click.getButton());

                return;
            }

            if (isHovered(mouseX, mouseY) && click.equals(Click.LEFT)) {
                // Set listening
                listening = !listening;

                SettingUpdateEvent settingUpdateEvent = new SettingUpdateEvent(setting);
                Paragon.INSTANCE.getEventBus().post(settingUpdateEvent);
            }

            super.mouseClicked(mouseX, mouseY, click);
        }
    }

    @Override
    public void mouseReleased(int mouseX, int mouseY, Click click) {
        if (setting.isVisible()) {
            super.mouseReleased(mouseX, mouseY, click);
        }
    }

    @Override
    public void keyTyped(int keyCode, char keyChar) {
        if (setting.isVisible()) {
            if (listening) {
                listening = false;

                setting.getValue().setDevice(Bind.Device.KEYBOARD);

                if (keyCode == Keyboard.KEY_DELETE || keyCode == Keyboard.KEY_BACK) {
                    setting.getValue().setButtonCode(0);
                    return;
                }

                setting.getValue().setButtonCode(keyCode);
            }
        }

        super.keyTyped(keyCode, keyChar);
    }

    @Override
    public float getHeight() {
        return getSetting().isVisible() ? super.getHeight() : 0;
    }

    @Override
    public float getSubElementsHeight() {
        return getSetting().isVisible() ? super.getSubElementsHeight() : 0;
    }

    @Override
    public float getTotalHeight() {
        return getSetting().isVisible() ? super.getTotalHeight() : 0;
    }

    public Setting<Bind> getSetting() {
        return setting;
    }

    public float getHover() {
        return hover;
    }

}

File: src/main/kotlin/com/paragon/client/ui/configuration/paragon/setting/impl/BindElement.kt
package com.paragon.client.ui.configuration.paragon.setting.impl

import com.paragon.Paragon
import com.paragon.api.event.client.SettingUpdateEvent
import com.paragon.api.setting.Bind
import com.paragon.api.setting.Setting
import com.paragon.api.util.render.ColourUtil.fade
import com.paragon.api.util.render.ColourUtil.integrateAlpha
import com.paragon.api.util.render.RenderUtil
import com.paragon.api.util.render.font.FontUtil
import com.paragon.client.ui.configuration.paragon.module.ModuleElement
import com.paragon.client.ui.configuration.paragon.setting.SettingElement
import com.paragon.client.ui.util.Click
import me.surge.animation.Animation
import me.surge.animation.Easing
import net.minecraft.util.math.MathHelper
import org.lwjgl.input.Keyboard
import java.awt.Color

/**
 * @author Surge
 * @since 06/08/2022
 */
class BindElement(setting: Setting<Bind>, module: ModuleElement, x: Float, y: Float, width: Float, height: Float) : SettingElement<Bind>(setting, module, x, y, width, height) {

    private val listening: Animation = Animation({ 200f }, false, { Easing.LINEAR })

    override fun draw(mouseX: Float, mouseY: Float, mouseDelta: Int) {
        RenderUtil.drawRect(x, y, width, height, Color(53, 53, 74).fade(Color(64, 64, 92), hover.getAnimationFactor()).rgb)

        RenderUtil.scaleTo(x + 5, y + 7, 0f, 0.5, 0.5, 0.5) {
            if (hover.getAnimationFactor() > 0.5) {
                FontUtil.drawStringWithShadow(if (listening.state) "Listening" else "Not Listening", x + 5, y + 7 + (7 * hover.getAnimationFactor()).toFloat(), Color(255, 0, 0).fade(Color(0, 255, 0), listening.getAnimationFactor()).integrateAlpha(MathHelper.clamp(255 * hover.getAnimationFactor(), 5.0, 255.0).toFloat()).rgb)
            }
        }

        RenderUtil.scaleTo(x + width - FontUtil.getStringWidth(setting.value.getButtonName()), y + 3, 0f, 0.7, 0.7, 0.7) {
            val side: Float = (x + width - 9 - FontUtil.getStringWidth(setting.value.getButtonName()) * 0.7f)

            FontUtil.drawStringWithShadow(setting.value.getButtonName(), side, y + 5, Color.GRAY.brighter().fade(Color.GRAY.brighter().brighter(), hover.getAnimationFactor()).rgb)
        }

        RenderUtil.scaleTo(x + 5, y + 5, 0f, 0.7, 0.7, 0.7) {
            FontUtil.drawStringWithShadow(setting.name, x + 5, y + 5 - (3 * hover.getAnimationFactor()).toFloat(), Color.GRAY.brighter().fade(Color.WHITE, listening.getAnimationFactor()).rgb)
        }

        super.draw(mouseX, mouseY, mouseDelta)
    }

    override fun mouseClicked(mouseX: Float, mouseY: Float, click: Click) {
        super.mouseClicked(mouseX, mouseY, click)

        if (click == Click.LEFT && hover.state) {
            listening.state = !listening.state
            return
        }

        if (listening.state) {
            listening.state = false

            setting.value.buttonCode = click.button
            setting.value.device = Bind.Device.MOUSE
            Paragon.INSTANCE.eventBus.post(SettingUpdateEvent(setting))
        }
    }

    override fun keyTyped(character: Char, keyCode: Int) {
        super.keyTyped(character, keyCode)

        if (listening.state) {
            listening.state = false

            setting.value.buttonCode = if (keyCode == Keyboard.KEY_DELETE || keyCode == Keyboard.KEY_BACK) 0 else keyCode
            setting.value.device = Bind.Device.KEYBOARD
            Paragon.INSTANCE.eventBus.post(SettingUpdateEvent(setting))
        }
    }

}